// backfill/lib/api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.put = exports.fetch = exports.computeHashOfOutput = exports.computeHash = exports.makeLogger = void 0;
const os_1 = require("os");
const backfill_config_1 = require("backfill-config");
const backfill_logger_1 = require("backfill-logger");
const backfill_hasher_1 = require("backfill-hasher");
const backfill_cache_1 = require("backfill-cache");
function makeConsole(stdout, stderr) {
    return {
        info(...args) {
            stdout.write(args.join(" ") + os_1.EOL);
        },
        warn(...args) {
            stderr.write(args.join(" ") + os_1.EOL);
        },
        error(...args) {
            stderr.write(args.join(" ") + os_1.EOL);
        },
    };
}
function makeLogger(logLevel, stdout, stderr) {
    return backfill_logger_1.makeLogger(logLevel, { console: makeConsole(stdout, stderr) });
}
exports.makeLogger = makeLogger;
async function computeHash(cwd, logger, hashSalt, config) {
    if (!config) {
        config = backfill_config_1.createConfig(logger, cwd);
    }
    const { packageRoot } = config;
    const hasher = new backfill_hasher_1.Hasher({ packageRoot }, logger);
    const hash = await hasher.createPackageHash(hashSalt || "");
    return hash;
}
exports.computeHash = computeHash;
async function computeHashOfOutput(cwd, logger, config) {
    if (!config) {
        config = backfill_config_1.createConfig(logger, cwd);
    }
    const { packageRoot } = config;
    const hasher = new backfill_hasher_1.Hasher({ packageRoot }, logger);
    const hash = await hasher.hashOfOutput();
    return hash;
}
exports.computeHashOfOutput = computeHashOfOutput;
async function fetch(cwd, hash, logger, config) {
    if (!config) {
        config = backfill_config_1.createConfig(logger, cwd);
    }
    const { cacheStorageConfig, internalCacheFolder, incrementalCaching } = config;
    const cacheStorage = backfill_cache_1.getCacheStorageProvider(cacheStorageConfig, internalCacheFolder, logger, cwd, incrementalCaching);
    const fetch = await cacheStorage.fetch(hash);
    return fetch;
}
exports.fetch = fetch;
async function put(cwd, hash, logger, config) {
    if (!config) {
        config = backfill_config_1.createConfig(logger, cwd);
    }
    const { cacheStorageConfig, internalCacheFolder, outputGlob, incrementalCaching, } = config;
    const cacheStorage = backfill_cache_1.getCacheStorageProvider(cacheStorageConfig, internalCacheFolder, logger, cwd, incrementalCaching);
    await cacheStorage.put(hash, outputGlob);
}
exports.put = put;
