// lage/lib/logger/Logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Logger = void 0;
const LogLevel_1 = require("./LogLevel");
const NpmLogReporter_1 = require("./reporters/NpmLogReporter");
class Logger {
    constructor() {
        this.logs = [];
    }
    log(level, msg, data) {
        const entry = {
            timestamp: Date.now(),
            level,
            msg,
            data,
        };
        this.logs.push(entry);
        for (const reporter of Logger.reporters) {
            reporter.log(entry);
        }
    }
    info(msg, data) {
        this.log(LogLevel_1.LogLevel.info, msg, data);
    }
    warn(msg, data) {
        this.log(LogLevel_1.LogLevel.warn, msg, data);
    }
    error(msg, data) {
        this.log(LogLevel_1.LogLevel.error, msg, data);
    }
    verbose(msg, data) {
        this.log(LogLevel_1.LogLevel.verbose, msg, data);
    }
    silly(msg, data) {
        this.log(LogLevel_1.LogLevel.silly, msg, data);
    }
}
exports.Logger = Logger;
Logger.reporters = [new NpmLogReporter_1.NpmLogReporter({ logLevel: LogLevel_1.LogLevel.info })];
