// lage/lib/logger/TaskLogger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TaskLogger = void 0;
const Logger_1 = require("./Logger");
const readline_1 = require("readline");
class TaskLogger {
    constructor(pkg, task) {
        this.pkg = pkg;
        this.task = task;
        this.logger = new Logger_1.Logger();
    }
    log(level, msg, data) {
        this.logger.log(level, msg, Object.assign({ package: this.pkg, task: this.task }, data));
    }
    info(msg, data) {
        this.logger.info(msg, Object.assign({ package: this.pkg, task: this.task }, data));
    }
    warn(msg, data) {
        this.logger.warn(msg, Object.assign({ package: this.pkg, task: this.task }, data));
    }
    error(msg, data) {
        this.logger.error(msg, Object.assign({ package: this.pkg, task: this.task }, data));
    }
    verbose(msg, data) {
        this.logger.verbose(msg, Object.assign({ package: this.pkg, task: this.task }, data));
    }
    silly(msg, data) {
        this.logger.silly(msg, Object.assign({ package: this.pkg, task: this.task }, data));
    }
    stream(level, input, data) {
        const readline = (0, readline_1.createInterface)({
            input,
            crlfDelay: Infinity,
        });
        readline.on("line", (line) => this.log(level, line, data));
    }
    getLogs() {
        return this.logger.logs;
    }
}
exports.TaskLogger = TaskLogger;
