// backfill-cache/lib/CacheStorage.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CacheStorage = void 0;
const globby_1 = __importDefault(require("globby"));
const hashFile_1 = require("./hashFile");
// First key is the hash, second key is the file relative path
const savedHashes = new Map();
// contract: cwd should be absolute
// The return keys are relative path with posix file separators
async function getHashesFor(cwd) {
    const result = new Map();
    const allFiles = await globby_1.default(["**/*", "!node_modules"], { cwd });
    //globby returns relative path with posix file separator
    await Promise.all(allFiles.map(async (f) => {
        const hash = await hashFile_1.getFileHash(cwd, f);
        result.set(f, hash);
    }));
    return result;
}
class CacheStorage {
    constructor(logger, cwd, incrementalCaching = false) {
        this.logger = logger;
        this.cwd = cwd;
        this.incrementalCaching = incrementalCaching;
    }
    async fetch(hash) {
        const tracer = this.logger.setTime("fetchTime");
        const result = await this._fetch(hash);
        tracer.stop();
        this.logger.setHit(result);
        if (!result && this.incrementalCaching) {
            savedHashes.set(hash, await getHashesFor(this.cwd));
        }
        return result;
    }
    async put(hash, outputGlob) {
        const tracer = this.logger.setTime("putTime");
        const filesMatchingOutputGlob = await globby_1.default(outputGlob, { cwd: this.cwd });
        let filesToCache = filesMatchingOutputGlob;
        if (this.incrementalCaching) {
            // Get the list of files that have not changed so we don't need to cache them.
            const hashesNow = await getHashesFor(this.cwd);
            const hashesThen = (await savedHashes.get(hash)) || new Map();
            const unchangedFiles = [...hashesThen.keys()].filter((s) => hashesThen.get(s) === hashesNow.get(s));
            filesToCache = filesMatchingOutputGlob.filter((f) => !unchangedFiles.includes(f));
        }
        await this._put(hash, filesToCache);
        tracer.stop();
    }
}
exports.CacheStorage = CacheStorage;
