// backfill-cache/lib/NpmCacheStorage.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NpmCacheStorage = void 0;
const path_1 = __importDefault(require("path"));
const execa_1 = __importDefault(require("execa"));
const fs_extra_1 = __importDefault(require("fs-extra"));
const globby_1 = __importDefault(require("globby"));
const CacheStorage_1 = require("./CacheStorage");
class NpmCacheStorage extends CacheStorage_1.CacheStorage {
    constructor(options, internalCacheFolder, logger, cwd, incrementalCaching = false) {
        super(logger, cwd, incrementalCaching);
        this.options = options;
        this.internalCacheFolder = internalCacheFolder;
    }
    async _fetch(hash) {
        const { npmPackageName, registryUrl, npmrcUserconfig } = this.options;
        const temporaryNpmOutputFolder = path_1.default.resolve(this.cwd, this.internalCacheFolder, "npm", hash);
        const packageFolderInTemporaryFolder = path_1.default.join(temporaryNpmOutputFolder, "node_modules", npmPackageName);
        if (!fs_extra_1.default.existsSync(packageFolderInTemporaryFolder)) {
            fs_extra_1.default.mkdirpSync(temporaryNpmOutputFolder);
            try {
                const runner = execa_1.default("npm", [
                    "install",
                    "--prefix",
                    temporaryNpmOutputFolder,
                    `${npmPackageName}@0.0.0-${hash}`,
                    "--registry",
                    registryUrl,
                    "--prefer-offline",
                    "--ignore-scripts",
                    "--no-shrinkwrap",
                    "--no-package-lock",
                    "--loglevel",
                    "error",
                    ...(npmrcUserconfig ? ["--userconfig", npmrcUserconfig] : []),
                ]);
                this.logger.pipeProcessOutput(runner.stdout, runner.stderr);
                await runner;
            }
            catch (error) {
                fs_extra_1.default.removeSync(temporaryNpmOutputFolder);
                if (error.stderr.toString().indexOf("ETARGET") > -1) {
                    return false;
                }
                else {
                    throw new Error(error);
                }
            }
        }
        const files = await globby_1.default(`**/*`, {
            cwd: packageFolderInTemporaryFolder,
        });
        await Promise.all(files.map(async (file) => {
            await fs_extra_1.default.mkdirp(path_1.default.dirname(path_1.default.join(this.cwd, file)));
            await fs_extra_1.default.copy(path_1.default.join(packageFolderInTemporaryFolder, file), path_1.default.join(this.cwd, file));
        }));
        return true;
    }
    async _put(hash, filesToCache) {
        const { npmPackageName, registryUrl, npmrcUserconfig } = this.options;
        const temporaryNpmOutputFolder = path_1.default.resolve(this.cwd, this.internalCacheFolder, "npm", hash, "upload");
        // Create package.json file
        fs_extra_1.default.outputJSONSync(path_1.default.join(temporaryNpmOutputFolder, "package.json"), {
            name: npmPackageName,
            version: `0.0.0-${hash}`,
        });
        await fs_extra_1.default.mkdirp(temporaryNpmOutputFolder);
        await Promise.all(filesToCache.map(async (file) => {
            const destinationFolder = path_1.default.join(temporaryNpmOutputFolder, path_1.default.dirname(file));
            await fs_extra_1.default.mkdirp(destinationFolder);
            await fs_extra_1.default.copy(path_1.default.join(this.cwd, file), path_1.default.join(temporaryNpmOutputFolder, file));
        }));
        // Upload package
        try {
            const runner = execa_1.default("npm", [
                "publish",
                "--registry",
                registryUrl,
                "--loglevel",
                "error",
                ...(npmrcUserconfig ? ["--userconfig", npmrcUserconfig] : []),
            ], {
                cwd: temporaryNpmOutputFolder,
                stdout: "inherit",
            });
            this.logger.pipeProcessOutput(runner.stdout, runner.stderr);
            await runner;
        }
        catch (error) {
            if (error.stderr.toString().indexOf("403") === -1) {
                throw new Error(error);
            }
        }
    }
}
exports.NpmCacheStorage = NpmCacheStorage;
