// backfill-hasher/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Hasher = exports.addToQueue = void 0;
const workspace_tools_1 = require("workspace-tools");
const hashOfFiles_1 = require("./hashOfFiles");
const hashOfPackage_1 = require("./hashOfPackage");
const helpers_1 = require("./helpers");
const repoInfo_1 = require("./repoInfo");
function isDone(done, packageName) {
    return Boolean(done.find(({ name }) => name === packageName));
}
function isInQueue(queue, packagePath) {
    return queue.indexOf(packagePath) >= 0;
}
function addToQueue(dependencyNames, queue, done, workspaces) {
    dependencyNames.forEach((name) => {
        const dependencyPath = workspace_tools_1.findWorkspacePath(workspaces, name);
        if (dependencyPath) {
            if (!isDone(done, name) && !isInQueue(queue, dependencyPath)) {
                queue.push(dependencyPath);
            }
        }
    });
}
exports.addToQueue = addToQueue;
class Hasher {
    constructor(options, logger) {
        this.options = options;
        this.logger = logger;
        this.packageRoot = this.options.packageRoot;
    }
    async createPackageHash(salt) {
        const tracer = this.logger.setTime("hashTime");
        const packageRoot = await helpers_1.getPackageRoot(this.packageRoot);
        this.repoInfo = await repoInfo_1.getRepoInfo(packageRoot);
        const { workspaceInfo } = this.repoInfo;
        const queue = [packageRoot];
        const done = [];
        while (queue.length > 0) {
            const packageRoot = queue.shift();
            if (!packageRoot) {
                continue;
            }
            const packageHash = await hashOfPackage_1.getPackageHash(packageRoot, this.repoInfo, this.logger);
            addToQueue(packageHash.internalDependencies, queue, done, workspaceInfo);
            done.push(packageHash);
        }
        const internalPackagesHash = hashOfPackage_1.generateHashOfInternalPackages(done);
        const buildCommandHash = helpers_1.hashStrings(salt);
        const combinedHash = helpers_1.hashStrings([internalPackagesHash, buildCommandHash]);
        this.logger.verbose(`Hash of internal packages: ${internalPackagesHash}`);
        this.logger.verbose(`Hash of build command: ${buildCommandHash}`);
        this.logger.verbose(`Combined hash: ${combinedHash}`);
        tracer.stop();
        this.logger.setHash(combinedHash);
        return combinedHash;
    }
    /**
     * Hash of output will hash the output files. This is meant to be used by validation and will not cache the repo hashes.
     * The validateOutput option should be used sparingly for performance reasons. It is meant to help be a debugging tool
     * to help investigate integrity of the cache.
     */
    async hashOfOutput() {
        const repoInfo = await repoInfo_1.getRepoInfoNoCache(this.packageRoot);
        return hashOfFiles_1.generateHashOfFiles(this.packageRoot, repoInfo);
    }
}
exports.Hasher = Hasher;
__exportStar(require("./repoInfo"), exports);
