// backfill-logger/lib/consoleLogger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeConsoleLogger = void 0;
const console_1 = require("./console");
const outputFormatter_1 = require("./outputFormatter");
const outputFilter_1 = require("./outputFilter");
function makeConsoleLogger(logLevel, overrides) {
    let consoleOverride = (overrides && overrides.console) || console_1.defaultConsole;
    let formatter = outputFormatter_1.defaultFormatter;
    let logFilter = outputFilter_1.defaultLogFilter(logLevel);
    return {
        consoleOverride,
        silly(...args) {
            if (logFilter.shouldLog("silly")) {
                consoleOverride.info(...formatter.format("silly", ...args));
            }
        },
        verbose(...args) {
            if (logFilter.shouldLog("verbose")) {
                consoleOverride.info(...formatter.format("verbose", ...args));
            }
        },
        info(...args) {
            if (logFilter.shouldLog("info")) {
                consoleOverride.info(...formatter.format("info", ...args));
            }
        },
        warn(...args) {
            if (logFilter.shouldLog("warn")) {
                consoleOverride.warn(...formatter.format("warn", ...args));
            }
        },
        error(...args) {
            if (logFilter.shouldLog("error")) {
                consoleOverride.error(...formatter.format("error", ...args));
            }
        },
        trace(...args) {
            if (logFilter.shouldLog("verbose")) {
                consoleOverride.error(...formatter.format("trace", ...args));
            }
        },
    };
}
exports.makeConsoleLogger = makeConsoleLogger;
