// backfill-logger/lib/logger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeLogger = exports.isCorrectLogLevel = void 0;
const path_1 = __importDefault(require("path"));
const fs_extra_1 = __importDefault(require("fs-extra"));
const filenamify_1 = __importDefault(require("filenamify"));
const chalk_1 = __importDefault(require("chalk"));
const timer_1 = require("./timer");
const consoleLogger_1 = require("./consoleLogger");
var logLevel_1 = require("./logLevel");
Object.defineProperty(exports, "isCorrectLogLevel", { enumerable: true, get: function () { return logLevel_1.isCorrectLogLevel; } });
function createFileName(performanceReportData) {
    return filenamify_1.default(`perf-${performanceReportData.name}-${performanceReportData.timestamp}.json`);
}
function makeLogger(logLevel, overrides) {
    const consoleLogger = consoleLogger_1.makeConsoleLogger(logLevel, overrides);
    const timer = timer_1.defaultTimer;
    const performanceReportData = {
        timestamp: Date.now(),
    };
    return {
        pipeProcessOutput(stdout, stderr) {
            stdout &&
                stdout.on("data", (chunk) => consoleLogger.consoleOverride.info(chunk.toString()));
            stderr &&
                stderr.on("data", (chunk) => consoleLogger.consoleOverride.error(chunk.toString()));
        },
        silly: consoleLogger.silly,
        verbose: consoleLogger.verbose,
        info: consoleLogger.info,
        warn: consoleLogger.warn,
        error: consoleLogger.error,
        setName(name) {
            consoleLogger.info(`Package name: ${name}`);
            performanceReportData["name"] = name;
        },
        setHash(hash) {
            consoleLogger.verbose(`Package hash: ${hash}`);
            performanceReportData["hash"] = hash;
        },
        setCacheProvider(cacheProvider) {
            consoleLogger.verbose(`Cache provider: ${cacheProvider}`);
            performanceReportData["cacheProvider"] = cacheProvider;
        },
        setHit(hit) {
            consoleLogger.info(hit ? `Cache hit!` : `Cache miss!`);
            performanceReportData["hit"] = hit;
        },
        setTime(type) {
            const tracer = timer.start();
            return {
                stop: () => {
                    const time = tracer.stop();
                    consoleLogger.verbose(`Profiling ${chalk_1.default.underline(type)} took ${chalk_1.default.cyanBright(`${time} ms`)}`);
                    performanceReportData[type] = time;
                },
            };
        },
        setMode(mode, logLevel) {
            consoleLogger[logLevel](`Running in ${mode} mode.`);
            performanceReportData["mode"] = mode;
        },
        setHashOfOutput(hash) {
            consoleLogger.verbose(`Hash of output: ${hash}`);
            performanceReportData["hashOfOutput"] = hash;
        },
        async toFile(logFolder) {
            const filepath = path_1.default.join(logFolder, createFileName(performanceReportData));
            await fs_extra_1.default.outputJson(filepath, performanceReportData, { spaces: 2 });
            consoleLogger.silly(`Performance log created at ${filepath}`);
        },
    };
}
exports.makeLogger = makeLogger;
