// backfill-config/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createConfig = exports.createDefaultConfig = exports.getSearchPaths = exports.getName = exports.isCorrectMode = exports.modesObject = void 0;
const path_1 = __importDefault(require("path"));
const pkg_dir_1 = __importDefault(require("pkg-dir"));
const find_up_1 = __importDefault(require("find-up"));
const envConfig_1 = require("./envConfig");
__exportStar(require("./cacheConfig"), exports);
__exportStar(require("./envConfig"), exports);
exports.modesObject = {
    READ_ONLY: "",
    WRITE_ONLY: "",
    READ_WRITE: "",
    PASS: "",
};
function isCorrectMode(mode) {
    return exports.modesObject.hasOwnProperty(mode);
}
exports.isCorrectMode = isCorrectMode;
function getName(packageRoot) {
    return (require(path_1.default.join(packageRoot, "package.json")).name ||
        path_1.default.basename(path_1.default.dirname(packageRoot)));
}
exports.getName = getName;
function getSearchPaths(fromPath) {
    const searchPaths = [];
    let nextPath = fromPath;
    while (nextPath) {
        const configLocation = find_up_1.default.sync("backfill.config.js", { cwd: nextPath });
        if (configLocation) {
            searchPaths.push(configLocation);
            nextPath = path_1.default.join(path_1.default.dirname(configLocation), "..");
        }
        else {
            nextPath = undefined;
        }
    }
    return searchPaths.reverse();
}
exports.getSearchPaths = getSearchPaths;
function createDefaultConfig(fromPath) {
    const packageRoot = pkg_dir_1.default.sync(fromPath) || fromPath;
    const defaultCacheFolder = path_1.default.join(packageRoot, "node_modules", ".cache", "backfill");
    const outputGlob = ["lib/**"];
    return {
        cacheStorageConfig: {
            provider: "local",
        },
        clearOutput: false,
        internalCacheFolder: defaultCacheFolder,
        logFolder: defaultCacheFolder,
        logLevel: "info",
        name: getName(packageRoot),
        mode: "READ_WRITE",
        outputGlob,
        packageRoot,
        producePerformanceLogs: false,
        validateOutput: false,
        incrementalCaching: false,
    };
}
exports.createDefaultConfig = createDefaultConfig;
function createConfig(logger, fromPath) {
    const defaultConfig = createDefaultConfig(fromPath);
    const fileBasedConfig = getSearchPaths(fromPath).reduce((acc, configPath) => {
        const config = require(configPath);
        if (config["mode"]) {
            if (!isCorrectMode(config["mode"])) {
                throw `Backfill config option "mode" was set, but with the wrong value: "${config["mode"]}".`;
            }
        }
        return {
            ...acc,
            ...config,
        };
    }, {});
    const envBasedConfig = envConfig_1.getEnvConfig(logger);
    return {
        ...defaultConfig,
        ...fileBasedConfig,
        ...envBasedConfig,
    };
}
exports.createConfig = createConfig;
